clippy_lints/src/invalid_build_cfg.rs

```diff
@@ -0,0 +1,308 @@
+                } else {
+                    format!(
+                        "{not}{open_paren}{cond}{closing_paren}",
+                        not = if parent_is_not { "!" } else { "" },
+                        open_paren = if !parent_is_not && is_top_level { "" } else { "(" },
+                        cond = v
+                            .iter()
+                            .map(|i| i.generate_replacement_inner(false, false))
+                            .collect::<Vec<_>>()
+                            .join(" || "),
+                        closing_paren = if !parent_is_not && is_top_level { "" } else { ")" },
+                    )
+                }
+            },
+            Self::All(v) => {
+                if v.is_empty() {
+                    if parent_is_not { "false" } else { "true" }.to_string()
+                } else if v.len() == 1 {
+                    v[0].generate_replacement_inner(is_top_level, parent_is_not)
+                } else {
+                    format!(
+                        "{not}{open_paren}{cond}{closing_paren}",
+                        not = if parent_is_not { "!" } else { "" },
+                        open_paren = if !parent_is_not && is_top_level { "" } else { "(" },
+                        cond = v
+                            .iter()
+                            .map(|i| i.generate_replacement_inner(false, false))
+                            .collect::<Vec<_>>()
+                            .join(" && "),
+                        closing_paren = if !parent_is_not && is_top_level { "" } else { ")" },
+                    )
+                }
+            },
+            Self::Not(i) => i.generate_replacement_inner(is_top_level, true),
+            Self::Feature(s) => format!(
+                "cfg!({}feature = {s}{})",
+                if parent_is_not { "not(" } else { "" },
+                if parent_is_not { ")" } else { "" },
+            ),
+        }
+    }
+}
+
+fn parse_macro_args(tokens: &TokenStream, has_unknown: &mut bool, out: &mut Vec<CfgAst>) {
+    let mut tokens = tokens.trees().peekable();
+
+    while let Some(token) = tokens.next() {
+        match token {
+            TokenTree::Token(token, _) => {
+                match token.kind {
+                    TokenKind::Ident(name, _) => {
+                        if name == sym::feature || name.as_str().starts_with("target_") {
+                            if let Some(next_token) = tokens.next()
+                                && let TokenTree::Token(next_token, _) = next_token
+                                && matches!(next_token.kind, TokenKind::Eq)
+                                && let Some(next_token) = tokens.next()
+                                && let TokenTree::Token(next_token, _) = next_token
+                                && let TokenKind::Literal(lit) = next_token.kind
+                                && matches!(lit.kind, LitKind::Str | LitKind::StrRaw(_))
```

Comment:
Instead of all this custom parsing logic could it be possible to use [`MetaItem::from_tokens`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/struct.MetaItem.html#method.from_tokens)? which is what I think the compiler is using internally; and then use `meta_list_item`.